New module to search a Signature (node): Issue #111

dongbohu opened the pull request:
This PR implements the new feature requested in issue #111.

dongbohu commented:
Live demo: http://165.123.67.202/#/node/search

You can ignore the two new issues reported by codeclimate.

rzelayafavila commented on interface/src/app/node/node_search.js:
```diff
@@ -0,0 +1,64 @@
+  });
+}])
+
+.factory('Node', ['$resource', 'ApiBasePath',
```
I think there is some repetition between the code here and the 'NodeInfo' factory that already exists in the ``node.js`` file. Should we consolidate them into one?

dongbohu replied:
It is true. But the URLs of the two resource classes are slightly different, so I didn't bother.

rzelayafavila replied:
Hmm are they really different though? It looks like the NodeInfo one just has the ``:id`` parameter for requests that are non-GET. It also looks like it does not use the ApiBasePath, which it probably should, right? Wouldn't it make sense to just keep one of them?

dongbohu replied:
I got some idea how to do it. Let me try.

rzelayafavila commented on interface/src/app/node/node_search.js:
```diff
@@ -0,0 +1,64 @@
+
+.config(['$stateProvider', function($stateProvider) {
+  $stateProvider.state('node_search', {
+    url: '/node/search',
```
Should this 'node_search' state be a child state of the 'node' state that already exists? What are the advantages/disadvantages?

dongbohu replied:
My original plan was to create an abstract state as parent and two child states for node information and node search. But I gave up for two reasons:

1) The two states "node" and "node_search" do not really share anything, so a parent state doesn't seem to be helpful. (I feel that their relationship is similar to the one between `gene_search` and `gene_network`.)

2) The file `node.js` has been pretty long now (475 lines). Adding the new `node_search` state to it will make the file longer. Keeping them separate will make it easier to maintain.

Does this make sense?

rzelayafavila replied:
Yeah, I think that is sensible. I think that the code for it wouldn't necessarily need to go in the ``node.js`` file, but if they do not share much in common and it would not be helpful then it makes sense not to do it.

rzelayafavila commented on interface/src/app/node/node_search.js:
```diff
@@ -0,0 +1,64 @@
+        function success(response) {
+          $scope.nodes = response.objects;
+          $scope.nodes.sort(function(n1, n2) { // Sort nodes by name
+            return (n1.name < n2.name) ? -1 : (n1.name === n2.name) ? 0 : 1;
```
A lot seems to be going on in this line, since we are using two ternary operators - I would suggest either breaking it up or adding a comment above it explain what the logic is.

dongbohu replied:
Sure. (Back to the good old time when I was a C/C++ programmer, I used this operator a lot. You probably can still see its popularity from today's linux kernel source code.)

rzelayafavila replied:
Yeah, I see and use ternary operators every now and then in JavaScript as well. I think that the chaining of the two operators together makes it a little harder to read, thus making a comment here helpful.

rzelayafavila commented on interface/src/index.html:
```diff
@@ -82,6 +82,11 @@
             </li>
+            <li ng-class="{active: inNodeStates()}">
+              <a ui-sref="node_search">
+                Signature
```
:+1:

rzelayafavila commented on interface/src/app/node/node_search.js:
```diff
@@ -0,0 +1,64 @@
+ * "adage.node_search" module.
+ */
+
+angular.module('adage.node_search', [
```
Quick question - should we use underscore when naming this module rather than camelCase? Do camelCase rules in javascript apply to module names as well?

dongbohu replied:
I don't know. Let me do some research and get back to you.

dongbohu commented:
@rzelayafavila: I think I addressed all your comments. I did some research and it seems that most angular developers prefer the module name in lowerCamelCase, so I renamed the module name from "node_search" to "nodeSearch".

In the new "resources.js", the main reason why Node resource service is named "NodeInfo" instead of "Node" is because `sampleBin.js` (written by @mhuyck) is referring to it. I am going to create a new issue to remind ourselves that. When his current PR is done, one of us can change the service name into "Node", and probably also move `NodeInfoSet` service into `node/resources.js` too.
